SDIOBlockDevice.h

```diff
@@ -0,0 +1,143 @@
+     */
+    virtual int deinit();
+
+    /** Read blocks from a block device
+     *
+     *  @param buffer   Buffer to write blocks to
+     *  @param addr     Address of block to begin reading from
+     *  @param size     Size to read in bytes, must be a multiple of read block size
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual int read(void *buffer, bd_addr_t addr, bd_size_t size);
+
+    /** Program blocks to a block device
+     *
+     *  The blocks must have been erased prior to being programmed
+     *
+     *  @param buffer   Buffer of data to write to blocks
+     *  @param addr     Address of block to begin writing to
+     *  @param size     Size to write in bytes, must be a multiple of program block size
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual int program(const void *buffer, bd_addr_t addr, bd_size_t size);
+
+    /** Mark blocks as no longer in use
+     *
+     *  This function provides a hint to the underlying block device that a region of blocks
+     *  is no longer in use and may be erased without side effects. Erase must still be called
+     *  before programming, but trimming allows flash-translation-layers to schedule erases when
+     *  the device is not busy.
+     *
+     *  @param addr     Address of block to mark as unused
+     *  @param size     Size to mark as unused in bytes, must be a multiple of erase block size
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual int trim(bd_addr_t addr, bd_size_t size);
+
+    /** Get the size of a readable block
+     *
+     *  @return         Size of a readable block in bytes
+     */
+    virtual bd_size_t get_read_size() const;
+
+    /** Get the size of a programable block
+     *
+     *  @return         Size of a programable block in bytes
+     *  @note Must be a multiple of the read size
+     */
+    virtual bd_size_t get_program_size() const;
+
+    /** Get the total size of the underlying device
+     *
+     *  @return         Size of the underlying device in bytes
+     */
+    virtual bd_size_t size() const;
+
+    /** Enable or disable debugging
+     *
+     *  @param dbg        State of debugging
+     */
+    virtual void debug(bool dbg);
+
+    /** Set the transfer frequency
+     *
+     *  @param freq     Transfer frequency
+     *  @note Max frequency supported is 25MHZ
+     */
+    virtual int frequency(uint64_t freq) { return BD_ERROR_OK; };
+
+    /** check if SD is present
+     *
+     *  @note check physical present switch. Maybe not support by hardware, then function will always return true.
+     */
+    virtual bool isPresent(void);
+
+  private:
+    DigitalIn _cardDetect;
+    bool _is_initialized;
+    bd_size_t _block_size;
+    bd_size_t _erase_size;
+    bd_size_t _sectors;
+    uint32_t _init_ref_count;
+    SD_Cardinfo_t _cardInfo;
+    uint32_t _card_type;
+
+    PlatformMutex _mutex;
```

Comment:
Can there be other SDIOBlockDevice instances or other SDIO devices that will try to use the same DMA 'driver' ?
If so .. this per-instance mutex will not protect parallel use of the DMA 'driver' is it ok?

Reply:
In this case with STM32F4, SDIO is a singleton because there is only one SDIO in hardware. I have not checked how to avoid to instantiate two SDIOBlockdevices, that should not be possible. There are also global vars preventing the use of multiple instances.
In STM32F7, there are devices with 2 hardware units, I have used only the first one in that implementation. In F7, the name for the device is also different, that is confusing.
Another suggestion in another conversation was to add the resource names in the constructor. That would be a cleaner solution for devices with several units, but I could not finish it until now.
This blockdevice driver was not accepted for the mbed-os, it should seperate into a HAL and device layer. I agree, but when this should be done its worth to work on the better solution. The singleton approach works now for the most boards with a SDIO interface (there are only a few). For the future, a HAL SDIO is better because it may be used for RF devices also, not only SD cards.

Reply:
> Can there be other SDIOBlockDevice instances or other SDIO devices 

Limiting to single instance is good idea for now.